### PR TITLE
Add support for recovery tokens

### PR DESCRIPTION
**Pull request by babolivier**
Following https://github.com/cozy/cozy-home/pull/769

Support for 2FA recovery tokens.
When an user tries to authenticate using one of its recovery tokens, Cozy log the user in, invalidate the used token (by removing it from the known tokens), and sends a notification to the user reminding it on how much tokens are remaining.
When there last token is used, another notification is sent, advising the user to regenerate its tokens so new tokens can be used.

### CODE REVIEW

**Comment by frankrousseau** [comment]
@m4dz or @aenario could you review this PR?

File: server/middlewares/authentication.coffee [hunk 1]
```diff
@@ -50,19 +53,45 @@ module.exports.authenticate = (req, res, next) ->
...
-                        error.status = 401
-                        next error
-                    else
-                        res.status(200).send success: true
+                recoveryCodes = user.encryptedRecoveryCodes[0]
+                if parseInt(req.body.authcode) in recoveryCodes
+                    # Disabling recovery code
+                    index = recoveryCodes.indexOf(parseInt req.body.authcode)
+                    recoveryCodes.splice index, 1
+                    console.log "------" + recoveryCodes.length
```
Comment (m4dz) [inline server/middlewares/authentication.coffee]: Maybe we can safely remove the `console.log` call (except if it's intended for logging)

Comment (babolivier) [inline server/middlewares/authentication.coffee]: Well, sometimes I forgot some logs when cleaning up after dev :smile:

**Comment by m4dz** [comment]
Ok, sounds good to me :+1:

File: server/middlewares/authentication.coffee (hunk 1 above)
Comment (m4dz) [inline server/middlewares/authentication.coffee]: This is the story of my life, buddy :smile:

**Comment by babolivier** [comment]
Thanks, just pushed a small edit as suggested :smile:

**Comment by babolivier** [comment]
Git messed my fork's history. I'll recreate a PR tonight.

**Comment by frankrousseau** [comment]
I merged it. Thank you a lot @babolivier

**Comment by babolivier** [comment]
Awesome, thanks @frankrousseau

**Pull request closed without merging**